### PR TITLE
Dynamically decide whether to do a Perfect Hash Join

### PR DESCRIPTION
**Pull request by lnkuiper**
Now that we dynamically compute the min/max for hash join builds, we can use this information to dynamically decide whether to do a perfect hash join (rather than statically while planning queries). The perfect hash join should trigger a lot more often.

I've had to remove `test/sql/join/inner/perfect_hash_join.test` because it checks whether the query plan contains a perfect hash join. However, now that we decide this dynamically, it cannot be printed in `PhysicalHashJoin::ParamsToString()`, as it is decided later than the function is called.

I guess we'll have to trust that the perfect HJ is still being triggered (I'm just kidding; the regression tests will clearly show when this breaks—the perfect HJ should speed up TPC-H, possibly more).

### CODE REVIEW

File: src/include/duckdb/execution/operator/join/physical_hash_join.hpp [hunk 1]
```diff
@@ -52,8 +50,9 @@ class PhysicalHashJoin : public PhysicalComparisonJoin {
 
 	//! Duplicate eliminated types; only used for delim_joins (i.e. correlated subqueries)
 	vector<LogicalType> delim_types;
-	//! Used in perfect hash join
-	PerfectHashJoinStats perfect_join_statistics;
+
+	//! Join Keys statistics (optional)
+	vector<unique_ptr<BaseStatistics>> join_stats;
```
Comment (Mytherin) [inline src/include/duckdb/execution/operator/join/physical_hash_join.hpp:55]: Maybe I'm missing something but it seems like this is never appended to, which would result in never triggering the fallback?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great. I have just one question about the fallback which appears to not be triggered.

For the fallback we also have two other options:

* We can choose to always collect min/max stats for the HT if a perfect hash join would be possible, even if we cannot generate filters to push dynamically
* We can choose to scan the HT to figure out the min/max if the HT is small enough, like we do now for the IN filters

For the `EXPLAIN` output being missing - I've been wanting to add a run-time addition to the `EXPLAIN` output which is triggered for `EXPLAIN ANALYZE` for the dynamically generated table filters. This could fit into that as well. But that should be done in a separate PR.

**Comment by lnkuiper** [comment]
Thanks for the review! I've implemented your feedback and spent some time trying to prevent regressions from computing the min/max, and I think I've got it sorted.

**Comment by Mytherin** [comment]
Nice! Looks good - thanks.

**Pull request merged**